### PR TITLE
Automated updating the dockerhub-description whenever a new version is released

### LINKED ISSUE #5: Automate updating the dockerhub description
Automate the process of updating [dockerhub-description.md](https://github.com/valkey-io/valkey-container/blob/mainline/dockerhub-description.md) in the valkey-container repository.

### PR DESCRIPTION
**Pull request by Nikhil-Manglore**
This python script will update the docker hub description whenever a new version of valkey is released. To test this I ran the script with the version as 8.1.0-rc2 and environment as "prod".

As displayed below a pull request was created on `Valkey-Container`:  

Screenshot 2025-03-21 at 2 04 47 AM_______
   

Here is the updated diff for `Valkey-Container`: https://github.com/Nikhil-Manglore/valkey-container/commit/78a35842fda6be7cfbb9499dcdc59417a2f10f09

We're mainly focusing on the diff for the `valkey-container/dockerhub-description.md` file. The image below shows the automation after releasing 8.1.0-rc2.

Screenshot 2025-03-21 at 2 06 50 AM_______

Here is the updated workflow:

Screenshot 2025-03-21 at 3 35 46 PM

Resolves #5

### CODE REVIEW

File: scripts/automate_docker_description.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    if section_end == float('inf'):
+        section_end = len(content)
+
+    # Get the content until the section header
+    section_content_start = content.find("\n", section_start) + 1
+    
+    # Replace only the content between current section and next section
+    return (content[:section_content_start] + "\n" + new_content + "\n\n" + content[section_end:])
+
+def update_docker_description(json_file: str, docker_description_file: str) -> None:
```
Comment (vudiep411) [inline scripts/automate_docker_description.py:50]: What is this `json_file`? Can u make the name a bit more descriptive so it's easier to understand?

File: .github/workflows/update-docker.yml [hunk 2]
```diff
@@ -83,4 +125,13 @@ jobs:
           body: |
             This pull request updates the repository to version `${{ inputs.version }}`.
             Please review the changes and merge when appropriate.
+
+      - name: Update Docker Hub Description
```
Comment (vudiep411) [inline .github/workflows/update-docker.yml:129]: We've decided to move this step to this repo?

**Comment by vudiep411** [comment]
I would recommend creating unit testing for the python script since this can have many edge cases. One possible way is to set up test cases and create another step that would run pytest in github action. This unit test should be part of the CI when someone raises a PR.

File: .github/workflows/update-docker.yml (hunk 2 above)
Comment (roshkhatri) [reply .github/workflows/update-docker.yml:129]: Just adding the description file changes to the PR

**Comment by roshkhatri** [comment]
Good callout @vudiep411 

@Nikhil-Manglore We do not want to update the docker hub from here, just update the PR

File: .github/workflows/update-docker.yml [hunk 3]
```diff
@@ -62,17 +68,53 @@ jobs:
...
+          token: ${{ secrets.PAT_TOKEN }}
+
+      - name: Generate bashbrew output
+        id: generate-bashbrew
+        run: |
+          cd valkey-container
+          BASHBREW_SCRIPTS=/home/runner/work/_actions/docker-library/bashbrew/v0.1.12/scripts
+          "$BASHBREW_SCRIPTS/github-actions/generate.sh" > bashbrew_output.json
+
+      - name: Update Docker Description
```
Comment (roshkhatri) [inline .github/workflows/update-docker.yml:90]:
```suggestion
      - name: Update Docker Description File
```

File: scripts/automate_docker_description.py (hunk 1 above)
Comment (Nikhil-Manglore) [reply scripts/automate_docker_description.py:49]: the json_file contains the [bashbrew output](https://github.com/valkey-io/valkey-container/actions/runs/13888909666/job/38857506155). We use the bashbrew output to extract the necessary information such as the directory and the tags.

File: scripts/automate_docker_description.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    
+    if section_end == float('inf'):
+        section_end = len(content)
+
+    # Get the content until the section header
+    section_content_start = content.find("\n", section_start) + 1
+    
+    # Replace only the content between current section and next section
+    return (content[:section_content_start] + "\n" + new_content + "\n\n" + content[section_end:])
+
```
Comment (roshkhatri) [inline scripts/automate_docker_description.py:48]: This looks quite complex, can we create a template file for the description and just fill in the blanks?
dockerhub-description-template.md and create a new dockerhub-description.m?
I think would be easier to follow.

We could also add a comment in dockerhub-description.md saying this is auto generated and do not edit it manually.

File: scripts/automate_docker_description.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    # Update each section separately
+    content = update_section(content, "## Official releases", official_releases)
+    content = update_section(content, "## Release candidates", release_candidates)
+    content = update_section(content, "## Latest unstable", latest_unstable)
+
+    with open(docker_description_file, 'w') as f:
+        f.write(content)
+
+if __name__ == "__main__":
+    update_docker_description(sys.argv[1], sys.argv[2])
```
Comment (roshkhatri) [inline scripts/automate_docker_description.py:87]: Can you add Error handling for number of arguments? and corresponding error logs?

File: scripts/automate_docker_description.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    
+    # Replace only the content between current section and next section
+    return (content[:section_content_start] + "\n" + new_content + "\n\n" + content[section_end:])
+
+def update_docker_description(json_file: str, docker_description_file: str) -> None:
+    with open(json_file) as f:
+        data = json.load(f)
+
+    with open(docker_description_file, 'r') as f:
+        content = f.read()
```
Comment (roshkhatri) [inline scripts/automate_docker_description.py:54]: Can you add Error handling for number of opening files? and corresponding error logs?

File: scripts/automate_docker_description.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+import sys
+
+def clean_tag(tag: str) -> str:
+    if ":" in tag:
+        return tag.split(":", 1)[1]
+    return tag
+
+def format_tag_line(entry: dict) -> str:
+    raw_tags = entry['meta']['entries'][0]['tags']
+
```
Comment (roshkhatri) [inline scripts/automate_docker_description.py:11]: Can you add Error handling for change in json structure.? and corresponding error logs?
something like a try blcok?

**Review by roshkhatri — COMMENTED** [review]
We might want to take a different approach for creating the description file.

**Comment by roshkhatri** [comment]
Can you also explore to edit the aliases here for the latest and the major version?
https://github.com/roshkhatri/valkey-container/blob/58c923048945c7b532914396508cbf061bd43ac7/generate-stackbrew-library.sh#L4

RC of new version we should not add it or edit it.

**Comment by Nikhil-Manglore** [comment]
> Can you also explore to edit the aliases here for the latest and the major version? https://github.com/roshkhatri/valkey-container/blob/58c923048945c7b532914396508cbf061bd43ac7/generate-stackbrew-library.sh#L4
> 
> RC of new version we should not add it or edit it.
> 
> Also, this would need to be done before running the update.sh script

Will make a separate issue for this

File: .github/workflows/update-docker.yml [hunk 8]
```diff
@@ -62,17 +64,54 @@ jobs:
...
+          git pull origin mainline --rebase
+
       - name: Check for Changes
         run: |
-          if git diff --exit-code; then
+          cd valkey-container
+          if git diff --exit-code origin/mainline; then
             echo "No changes detected. Skipping PR creation."
             exit 0
           fi
```
Comment (vudiep411) [inline .github/workflows/update-docker.yml:107]: if there are no changes, would `peter-evans/create-pull-request@v7` just skip by itself or we still need this step?

Comment (Nikhil-Manglore) [reply .github/workflows/update-docker.yml:113]: Good catch, according to the documentation in `peter-evans/create-pull-request@v7` the action will not create a PR if there are no changes. This step is redundant, will remove it.

File: .github/workflows/update-docker.yml [hunk 9]
```diff
@@ -62,17 +64,54 @@ jobs:
...
+
+      - name: Ensure Up-to-Date Branch
+        run: |
+          cd valkey-container
+          git config --global user.email "github-actions[bot]@users.noreply.github.com"
+          git config --global user.name "github-actions[bot]"
+          git add .
+          git commit -m "Preparing changes for version ${{ inputs.version }}"
+          git checkout mainline
+          git pull origin mainline --rebase
```
[the remark this answers is not in the document]
Comment (Nikhil-Manglore) [reply .github/workflows/update-docker.yml:105]: You're correct, `peter-evans/create-pull-request@v7` will already handle all the necessary git operations, will remove this step entirely.

File: scripts/automate_docker_description.py [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+        with open(output_file, 'w') as f:
+            f.write(content)
+
+    except KeyError as e:
+        logging.error(f"Invalid JSON structure: {e}")
+        sys.exit(1)
+    except Exception as e:
+        logging.error(f"Error processing data: {e}")
+        sys.exit(1)
```
Comment (vudiep411) [inline scripts/automate_docker_description.py:91]: You could merge these into 1 try catch block instead of splitting them.
```python
def update_docker_description(json_file: str, template_file: str, output_file: str) -> None:
    try:
    # Your logic
    ...
    except FileNotFoundError as e:
        logging.error(f"File not found: {e}")
        sys.exit(1)
    except json.JSONDecodeError:
        logging.error(f"Failed to parse JSON file '{json_file}'. Please check its syntax.")
        sys.exit(1)
    except KeyError as e:
        logging.error(f"Invalid JSON structure: {e}")
        sys.exit(1)
    except Exception as e:
        logging.error(f"Error processing data: {e}")
        sys.exit(1)

File: .github/workflows/update-docker.yml [hunk 11]
```diff
@@ -61,26 +63,50 @@ jobs:
 
             # Run update script
             ./update.sh ${{ steps.strip_version.outputs.stripped_version }}
+            git config user.name "github-actions[bot]"
+            git config user.email "github-actions[bot]@users.noreply.github.com"
+            git add .
+            git commit -m "Update to version ${{ inputs.version }}"
```
Comment (vudiep411) [inline .github/workflows/update-docker.yml:69]: nit: If u want, u could probably create a new step for the commit step here like this
```yml
- name: Commit changes to generate bashbrew output
   run: |
          ...
```
it might be easier for other to read

File: documents/dockerhub-description-template.md [hunk 12]
```diff
@@ -0,0 +1,105 @@
+## THIS FILE IS AUTO-GENERATED - DO NOT EDIT MANUALLY
```
Comment (roshkhatri) [inline documents/dockerhub-description-template.md:1]: I would also add which file generates this file.

**Review by roshkhatri — COMMENTED** [review]
I would ideally move this file to valkey-container. Please open a PR for valkey-container and you you can use the template when we checkout the repository.

File: documents/dockerhub-description-template.md (hunk 12 above)
Comment (Nikhil-Manglore) [reply documents/dockerhub-description-template.md:1]: Just to clarify, you want the template file moved to valkey-container?

File: .github/workflows/update-docker.yml (hunk 11 above)
Comment (Nikhil-Manglore) [reply .github/workflows/update-docker.yml:69]: I think it's fine how it is

Comment (roshkhatri) [inline .github/workflows/update-docker.yml:69]: do we need need to sign off here?

File: .github/workflows/update-docker.yml [hunk 13]
```diff
@@ -61,26 +63,50 @@ jobs:
...
+          cd valkey-container
+          BASHBREW_SCRIPTS=/home/runner/work/_actions/docker-library/bashbrew/v0.1.12/scripts
+          "$BASHBREW_SCRIPTS/github-actions/generate.sh" > bashbrew_output.json
+
+      - name: Update Docker Description File
+        run: |
+          python valkey-release-automation/scripts/automate_docker_description.py valkey-container/bashbrew_output.json valkey-container/dockerhub-description-template.md valkey-container/dockerhub-description.md
+          rm valkey-container/bashbrew_output.json
+          cd valkey-container
+          git reset HEAD~1
```
Comment (roshkhatri) [inline .github/workflows/update-docker.yml:95]: We dont need to squash the commits

File: .github/workflows/update-docker.yml [hunk 14]
```diff
@@ -61,26 +63,50 @@ jobs:
...
+        uses: actions/setup-python@v4
+        with:
+          python-version: '3.x'
+
+      - name: Checkout valkey-release-automation
+        uses: actions/checkout@v4
+        with:
+          repository: ${{ github.repository_owner }}/valkey-release-automation
+          path: valkey-release-automation
+          token: ${{ secrets.PAT_TOKEN }}
```
Comment (roshkhatri) [inline .github/workflows/update-docker.yml:81]: we can remove this

File: .github/workflows/update-docker.yml (hunk 11 above)
Comment (roshkhatri) [inline .github/workflows/update-docker.yml:69]:
```suggestion
            git commit -m "Update versions.json and Dockerfiles for version ${{ inputs.version }}" -s
```

File: .github/workflows/update-docker.yml (hunk 13 above)
Comment (roshkhatri) [inline .github/workflows/update-docker.yml:95]:
```suggestion
         git add .
         git commit -m "Update tags in the docker description for version ${{ inputs.version }}" -s
```

**Review by roshkhatri — APPROVED** [review]
LGTM
